File: TopCoder/python/WordBreak.py
from typing import List

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        #res = self.recursive(0, s, wordDict)
        dp = [False] * (len(s) + 1)
        dp[0] = True

        map = {}
        for word in wordDict:
            len_word = len(word)
            list = map.get(len_word, [])
            list.append(word)
            map[len_word] = list

        index = 1
        while index <= len(s):
            for key, value in map.items():
                if key <= index:
                    suffix = index - key
                    sub_str = s[suffix: suffix + key]
                    if dp[suffix] and map.get(len(sub_str), None) and sub_str in map.get(len(sub_str)):
                        dp[index] = True
            index += 1
        return dp[len(dp) - 1]



    def recursive(self, index: int, s: str, wordDict: List[str]) -> bool:
        if index >= len(s):
            return True

        for word in wordDict:
            sub_str = s[index: index + len(word)]
            if sub_str == word:
                if self.recursive(index+len(word), s, wordDict):
                    return True

        return False


    def isPresent(self, s: str, wordDict: List[str]) -> bool:
        try:
            if wordDict.index(s, 0, len(wordDict)) >= 0:
                return True
        except ValueError:
            return False

if __name__ == '__main__':
    sol = Solution()
    res = sol.wordBreak("catsandog",["cats", "dog", "sand", "and", "cat"])
    #res = sol.wordBreak("leetcode", ["leet", "code"])
    print(res)

File: TopCoder/src/main/java/com/leetcode/String/WordBreak.java
package com.leetcode.String;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

//https://leetcode.com/problems/word-break/
public class WordBreak {

    public boolean wordBreak(String s, List<String> wordDict) {
        //Map<Integer, List<String>> map = wordDict.stream().collect(Collectors.groupingBy(String::length));
        //System.out.println(map);
        //boolean result = recursive(0, s, map);
        boolean result = dp(s, wordDict);
        return result;
    }

    private boolean recursive(int index, String input, Map<Integer, List<String>> wordDict) {
        if (index >= input.length()) {
            return true;
        }

        for (Map.Entry<Integer, List<String>> entry : wordDict.entrySet()) {
            String str = input.substring(index, index + entry.getKey());
            if (entry.getValue().contains(str) && recursive(index + entry.getKey(), input, wordDict)) {
                return true;
            }
        }
        return false;
    }

    private boolean dp(String s, List<String> wordDict) {
        boolean dp[] = new boolean[s.length()+1];
        dp[0] = true;
        Map<Integer, List<String>> map = wordDict.stream().collect(Collectors.groupingBy(String::length));
        for (int i = 1 ; i <= s.length() ; i++) {
            for (Map.Entry<Integer, List<String>> entry : map.entrySet()) {
                if (entry.getKey() <= (i)) {
                    int length2 = (i) - entry.getKey();
                    String lastSuffix = s.substring(length2, i);
                    if (dp[length2] && map.containsKey(lastSuffix.length()) && map.get(lastSuffix.length()).contains(lastSuffix)) {
                        dp[i] = true;
                        break;
                    }
                }
            }
        }
        return dp[dp.length - 1];
    }


    public static void main(String[] args) {
        String input = "catsandog";
        List<String> list = Arrays.asList("cats", "dog", "sand", "and", "cat");
        //String input = "applepenapple";
        //List<String> list = Arrays.asList("apple", "pen");
        boolean result = new WordBreak().wordBreak(input, list);
        System.out.println(result);
    }
}
